pubsub/democlient/main.go

```diff
@@ -142,84 +145,65 @@ func main() {
 					}
 				}
 				a.data = data
+
+				err := subscribe(a.data)
+				if err != nil {
+					fmt.Printf("Failed to subscribe: %v", err)
+					continue
+				}
+
 			case "unsubscribe":
 				unsubPrompt.Options = currentSubs
 				_ = survey.AskOne(unsubPrompt, &a.data, nil)
+
+				err := unsubscribe(a.data)
+				if err != nil {
+					fmt.Printf("Failed to unsubscribe: %v", err)
+					continue
+				}
+
 			case "quit":
-				close(promptAgain)
+				cancel()
 				os.Exit(0)
+
 			default:
 				log.Fatalf("invalid action")
 				continue
 			}
-
-			if len(a.data) == 0 {
-				//actionChan <- &actionData{"", nil}
-				go func() { promptAgain <- struct{}{} }()
-				continue
-			}
-
-			log.Printf("Submitting %s request...", a.action)
-			actionChan <- &a
 		}
 	}()
-	promptAgain <- struct{}{}
 
-	// Send/receive messages in an orderly fashion.
+	// Receive subscribed broadcast messages in an orderly fashion.
 	for {
-		select {
-		case a := <-actionChan:
-			switch a.action {
-			case "subscribe":
-				if err = subscribe(a.data); err != nil {
-					log.Fatalf("subscribed failed: %v", err)
-				}
-			case "unsubscribe":
-				if err = unsubscribe(a.data); err != nil {
-					log.Fatalf("subscribed failed: %v", err)
-				}
-			case "quit":
-				close(promptAgain)
-				close(actionChan)
-				os.Exit(0)
-			}
-
-			promptAgain <- struct{}{}
-		default:
-			//log.Println("No actions received. Going on to wait for messages.")
-		}
-
-		resp, err := cl.ReceiveMsg()
-		if err != nil {
-			if pstypes.IsIOTimeoutErr(err) {
-				continue
-			}
+		msg := <-cl.Receive()
+		if msg == nil {
 			fmt.Printf("ReceiveMsg failed: %v", err)
 			return
 		}
 
-		msg, err := client.DecodeMsg(resp)
-		if err != nil {
-			log.Printf("Failed to decode message: %v", err)
-			continue
-		}
-
-		switch m := msg.(type) {
+		switch m := msg.Message.(type) {
+		case *pstypes.ResponseMessage:
```

Comment:
Responses to explicit requests (e.g. subscribe, unsubscribe) get a ResponseMessage type in return.  Other message types are broadcast messages.